python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,294 @@
+    float to a 64-bit (double precision) float which introduces some imprecision.
+    """
+    return struct.pack("<f", value)
+
+
+@to_bytes.register(DoubleType)
+def _(primitive_type, value: float) -> bytes:
+    return struct.pack("<d", value)
+
+
+@to_bytes.register(StringType)
+def _(primitive_type, value: str) -> bytes:
+    return value.encode("UTF-8")
+
+
+@to_bytes.register(UUIDType)
+def _(primitive_type, value: uuid.UUID) -> bytes:
+    return struct.pack(">QQ", (value.int >> 64) & 0xFFFFFFFFFFFFFFFF, value.int & 0xFFFFFFFFFFFFFFFF)
+
+
+@to_bytes.register(BinaryType)
+@to_bytes.register(FixedType)
+def _(primitive_type, value: bytes) -> bytes:
+    return value
+
+
+@to_bytes.register(DecimalType)
+def _(primitive_type, value: Decimal) -> bytes:
+    unscaled_value = convert_decimal_to_unscaled_value(decimal_type_var=primitive_type, value=value)
+    min_num_bytes = ((unscaled_value).bit_length() + 7) // 8
+    return unscaled_value.to_bytes(min_num_bytes, "big", signed=True)
+
+
+@singledispatch
+def from_bytes(primitive_type: PrimitiveType, b: bytes) -> Union[bool, bytes, Decimal, float, int, str, uuid.UUID]:
+    """A generic function which converts bytes to a built-in python value
+
+    Args:
+        primitive_type(PrimitiveType): An implementation of the PrimitiveType base class
+        b(bytes): The bytes to convert
+    """
+    raise TypeError(f"Cannot deserialize bytes, type {primitive_type} not supported: {str(b)}")
+
+
+@from_bytes.register(BooleanType)
+def _(primitive_type, b: bytes) -> bool:
+    return struct.unpack("<?", b)[0] != 0
+
+
+@from_bytes.register(IntegerType)
+@from_bytes.register(DateType)
+def _(primitive_type, b: bytes) -> int:
+    return struct.unpack("<i", b)[0]
+
+
+@from_bytes.register(LongType)
+@from_bytes.register(TimeType)
+@from_bytes.register(TimestampType)
+@from_bytes.register(TimestamptzType)
+def _(primitive_type, b: bytes) -> int:
+    return struct.unpack("<q", b)[0]
+
+
+@from_bytes.register(FloatType)
+def _(primitive_type, b: bytes):
+    return struct.unpack("<f", b)[0]
+
+
+@from_bytes.register(DoubleType)
+def _(primitive_type, b: bytes) -> float:
+    return struct.unpack("<d", b)[0]
+
+
+@from_bytes.register(StringType)
+def _(primitive_type: PrimitiveType, b: bytes) -> str:
+    return bytes(b).decode("utf-8")
+
+
+@from_bytes.register(UUIDType)
+def _(primitive_type, b: bytes) -> uuid.UUID:
+    unpacked_bytes = struct.unpack(">QQ", b)
+    return uuid.UUID(int=unpacked_bytes[0] << 64 | unpacked_bytes[1])
+
+
+@from_bytes.register(BinaryType)
+@from_bytes.register(FixedType)
+def _(primitive_type, b: bytes) -> bytes:
+    return b
+
+
+@from_bytes.register(DecimalType)
+def _(primitive_type, b: bytes) -> Decimal:
+    integer_representation = int.from_bytes(b, "big", signed=True)
+    with localcontext() as ctx:
+        ctx.prec = primitive_type.precision
+        return Decimal(integer_representation) * Decimal((0, (1,), -primitive_type.scale))
```

Comment:
One more idea here: `Decimal(integer_representation)` creates a decimal with the right sign and digits. The right scale is `-primitive_type`. So we can avoid doing any decimal arithmetic (and hitting context problems) by using this:

```python
def unscaled_to_decimal(unscaled: int, scale: int) -> Decimal:
    sign, digits, _ = Decimal(unscaled).to_tuple()
    return Decimal((sign, digits, scale))

@from_bytes.register(DecimalType)
def _(decimal_type, buf: bytes) -> Decimal:
    unscaled = int.from_bytes(buf, "big", signed=True)
    return unscaled_to_decimal(unscaled, -primitive_type.exponent)
```

Reply:
Much cleaner! Switched to this and removed `localcontext` use entirely.